components/style/values/specified/basic_shape.rs

```diff
@@ -10,347 +10,263 @@
-fn serialize_basicshape_position<W>(position: &Position, dest: &mut W) -> fmt::Result
-    where W: fmt::Write
-{
-    fn to_keyword_and_lop<S>(component: &PositionComponent<S>) -> (S, Cow<LengthOrPercentage>)
-        where S: Copy + Side
-    {
-        match *component {
-            PositionComponent::Center => {
-                (S::start(), Cow::Owned(LengthOrPercentage::Percentage(Percentage(0.5))))
-            },
-            PositionComponent::Side(keyword, None) => {
-                // left | top => 0%
-                // right | bottom => 100%
-                let p = if keyword.is_start() { 0. } else { 1. };
-                (S::start(), Cow::Owned(LengthOrPercentage::Percentage(Percentage(p))))
-            },
-            PositionComponent::Side(keyword, Some(ref lop)) if !keyword.is_start() => {
-                if let LengthOrPercentage::Percentage(p) = *to_non_zero_length(lop) {
-                    (S::start(), Cow::Owned(LengthOrPercentage::Percentage(Percentage(1. - p.0))))
-                } else {
-                    (keyword, Cow::Borrowed(lop))
-                }
-            },
-            PositionComponent::Length(ref lop) |
-            PositionComponent::Side(_, Some(ref lop)) => {
-                (S::start(), to_non_zero_length(lop))
-            },
-        }
-    }
+        let mut widths = parse_one_set_of_border_values(context, input)?;
+        let mut heights = if input.try(|input| input.expect_delim('/')).is_ok() {
+            parse_one_set_of_border_values(context, input)?
+        } else {
+            [widths[0].clone(),
+             widths[1].clone(),
+             widths[2].clone(),
+             widths[3].clone()]
+        };
 
-    fn to_non_zero_length(lop: &LengthOrPercentage) -> Cow<LengthOrPercentage> {
-        match *lop {
-            LengthOrPercentage::Length(ref l) if l.is_zero() => {
-                Cow::Owned(LengthOrPercentage::Percentage(Percentage(0.)))
-            },
-            _ => {
-                Cow::Borrowed(lop)
-            }
-        }
+        Ok(BorderRadius {
+            top_left: BorderRadiusSize::new(widths[0].take(), heights[0].take()),
+            top_right: BorderRadiusSize::new(widths[1].take(), heights[1].take()),
+            bottom_right: BorderRadiusSize::new(widths[2].take(), heights[2].take()),
+            bottom_left: BorderRadiusSize::new(widths[3].take(), heights[3].take()),
+        })
     }
+}
 
-    fn write_pair<A, B, W>(a: &A, b: &B, dest: &mut W) -> fmt::Result
-        where A: ToCss, B: ToCss, W: fmt::Write
-    {
-        a.to_css(dest)?;
-        dest.write_str(" ")?;
-        b.to_css(dest)
-    }
+fn parse_one_set_of_border_values(context: &ParserContext, mut input: &mut Parser)
+                                 -> Result<[LengthOrPercentage; 4], ()> {
+    let a = try!(LengthOrPercentage::parse_non_negative(context, input));
+    let b = if let Ok(b) = input.try(|i| LengthOrPercentage::parse_non_negative(context, i)) {
```

Comment:
nit: Use `match` instead of `if let { ... } else { return }`?

Here and below. You save a line in some of those I think ;)